### PR TITLE
GuidedTours: Use JSX for main tour

### PR DESCRIPTION
**Pull request by ehg**
Split from #7030

Aims to be a more minimal PR, so I've stripped out:
 - analytics
 - blanket exit
 - positioning fixes
 - themes tour

There should be more to strip out, that we can put back in — in later PRs. I think we should concentrate on functional feature parity with the old approach first. And also reduce review overhead.

TODO:
- [x] Completely remove dependency on old guided tours config
- [x] Decide where to put `wait` code — *just delete it for now*
- [x] Update PropTypes for config elements
- [x] Tidy
- [x] Squash

To test:
- go to http://calypso.localhost:3000 and click around — just check whether anything is obviously broken
- go to http://calypso.localhost:3000/?tour=main — check whether you can proceed through the main tour (see below) properly
- also check whether exiting the tour (via the No Thanks / Quit buttons) works

GIF of the main tour as it should appear: 

main-tour

Test live: https://calypso.live/?branch=update/guidedtours-framework-to-jsx

### CODE REVIEW

File: client/layout/guided-tours/config-elements.js [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+		const { children, context, isValid, lastAction, stepName } = this.props;
+		const nextStep = find( children, stepComponent =>
+			stepComponent.props.name === stepName );
+		const isLastStep = nextStep === children[ children.length - 1 ];
+
+		if ( ! nextStep || ! isValid( context ) ) {
+			return null;
+		}
+
+		return React.cloneElement( nextStep, { isLastStep, isValid, lastAction } );
```
Comment (mcsf) [inline client/layout/guided-tours/config-elements.js]: `isValid` and `lastAction` are already carried in `context`, so they shouldn't be passed here. Likewise, `Step`'s methods should look for those properties in `context` rather than `props`.

Comment (ehg) [inline client/layout/guided-tours/config-elements.js]: ce746f6 addresses this, although we must be careful to use the updated context, to avoid getting stale `state` binds.

Any ideas on naming? :]

File: client/layout/guided-tours/main-tour.js [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+	previewIsNotShowing,
+	previewIsShowing,
+} from 'state/ui/guided-tours/contexts';
+import Gridicon from 'components/gridicon';
+
+const context = state =>
+	true || isNewUser( state );
+
+export const MainTour = makeTour(
+	<Tour name="main" version="20160601" path="/" context={ context }>
```
Comment (ehg) [inline client/layout/guided-tours/main-tour.js]: I wanted to do something like `context={ and( isNewUser, isEnabled( 'guided-tours/main' ) ) }` here, but `tour=main` doesn't force the tour then.

I *think* forcing a tour via a qarg should respect the `path`, but override the `context` of the tour. This way tours can be accessed by links, but won't break by being shown on an incorrect path.

**Comment by ehg** [comment]
Rebased for node 6.

File: client/layout/guided-tours/config-elements.js (hunk 1 above)
Comment (mcsf) [inline client/layout/guided-tours/config-elements.js]: Damned naming. Well, React's own context will always be `context`, so let's rethink our config attribute. Would `when` suffice? `condition`?

Comment (ehg) [inline client/layout/guided-tours/config-elements.js]: I think `when` would work nicely, especially for `Continue` steps. Will update.

**Comment by lsinger** [comment]
Currently I can qarg-force the `main` tour, but not the `themes` tour.

**Comment by ehg** [comment]
> Currently I can qarg-force the main tour, but not the themes tour.

The themes tour doesn't exist in this PR.

**Comment by lsinger** [comment]
> The themes tour doesn't exist in this PR.
:facepalm:

File: client/layout/guided-tours/main-tour.js (hunk 2 above)
Comment (ehg) [inline client/layout/guided-tours/main-tour.js]: Added 250462c & 0a8fd6d. We can probably defer respecting the `path` until a later PR.

File: client/layout/guided-tours/main-tour.js [hunk 3]
```diff
@@ -13,18 +13,16 @@ import {
...
 	previewIsShowing,
 } from 'state/ui/guided-tours/contexts';
 import Gridicon from 'components/gridicon';
 
-const context = state =>
-	true || isNewUser( state );
-
 export const MainTour = makeTour(
-	<Tour name="main" version="20160601" path="/" when={ context }>
+	<Tour name="main" version="20160601" path="/" when={ and( isNewUser, isEnabled( 'guided-tours/main' ) ) }>
```
Comment (mcsf) [inline client/layout/guided-tours/main-tour.js]: This makes me happy.

File: client/state/ui/guided-tours/actions.js [hunk 4]
```diff
@@ -46,7 +45,7 @@ export function nextGuidedTourStep( { tour, stepName } ) {
 
 	const trackEvent = recordTracksEvent( 'calypso_guided_tours_next_step', {
 		step: stepName,
-		tour_version: guidedToursConfig.get( tour ).meta.version,
+		//tour_version: guidedToursConfig.get( tour ).meta.version,
```
Comment (lsinger) [inline client/state/ui/guided-tours/actions.js]: Do we want to leave this as is for this PR? I remember seeing discussions on how to handle this, but didn't see the decision.

Comment (ehg) [inline client/state/ui/guided-tours/actions.js]: I think so, reworking analytics can be a separate PR.

File: client/layout/guided-tours/config-elements.js [hunk 5]
```diff
@@ -76,7 +79,13 @@ export class Tour extends Component {
 export class Step extends Component {
 	static propTypes = {
 		name: PropTypes.string.isRequired,
-	}
+		placement: PropTypes.string,
+		next: PropTypes.string,
+		target: PropTypes.string,
+		arrow: PropTypes.string,
```
Comment (ehg) [inline client/layout/guided-tours/config-elements.js]: D'you think arrow types would be useful here?

Comment (lsinger) [inline client/layout/guided-tours/config-elements.js]: Definitely. If only for documentation.

Comment (lsinger) [inline client/layout/guided-tours/config-elements.js]: Placement types as well, for that matter.

Comment (lsinger) [inline client/layout/guided-tours/config-elements.js]: OTOH `.oneOf( [ 'below', 'beside', ... ] )` would probably be sufficient in both cases.

Comment (lsinger) [inline client/layout/guided-tours/config-elements.js]: 62d80c8

File: client/layout/guided-tours/config-elements.js [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+const contextTypes = Object.freeze( {
+	next: PropTypes.func.isRequired,
+	quit: PropTypes.func.isRequired,
+	isValid: PropTypes.func.isRequired,
+	tour: PropTypes.string.isRequired,
+	tourVersion: PropTypes.string.isRequired,
+	step: PropTypes.string.isRequired,
+} );
+
+export class Tour extends Component {
```
Comment (bisko) [inline client/layout/guided-tours/config-elements.js]: Is there a reason to keep all the components in a single file, instead of organizing them in separate files for each component?

Comment (ehg) [inline client/layout/guided-tours/config-elements.js]: Ease of initial development atm. We'll likely split code up in the coming PRs, after this big-switch is merged.

**Comment by lsinger** [comment]
It does LGTM, but I'd prefer to have more eyes on it. @mcsf?

File: client/state/ui/guided-tours/actions.js (hunk 4 above)
Comment (mcsf) [inline client/state/ui/guided-tours/actions.js]: The line is still commented out, is that intentional?

**Comment by mcsf** [comment]
Left a question re analytics.

That aside: :shipit: . This is good work.

File: client/state/ui/guided-tours/actions.js (hunk 4 above)
Comment (mcsf) [inline client/state/ui/guided-tours/actions.js]: All sorted out. Carry on.

**Comment by lsinger** [comment]
💥 😸
